app/services/solidus_afterpay/update_order_attributes_service.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module SolidusAfterpay
+  class UpdateOrderAttributesService < BaseService
+    def initialize(order:, afterpay_order_token:, payment_method:, request_env:)
+      @order = order
+      @afterpay_order_token = afterpay_order_token
+      @payment_method = payment_method
+      @request_env = request_env
+
+      super()
+    end
+
+    def call
+      return false unless afterpay_order.success?
+
+      ::Spree::OrderUpdateAttributes.new(order, update_params, request_env: request_env).apply
+    end
+
+    private
+
+    def afterpay_order
+      @afterpay_order ||= payment_method.gateway.find_order(token: afterpay_order_token)
+    end
+
+    def shipping_rate
+      @shipping_rate ||= ::Spree::ShippingRate.find(afterpay_order.params.shippingOptionIdentifier.to_i)
+    end
+
+    def update_params
+      {
+        email: afterpay_order.params.consumer.email,
```

Comment:
I don't really like this stuff 😞 
The `payment_method.gateway.find_order(token: afterpay_order_token)` doesn't return an `afterpay_order` but an `ActiveMerchant::Billing::Response` so you need to use `afterpay_order.params` to access the real object. I still prefer an implementation like the old `find_payment` method 🤔